src/annotator/plugin/test/bucket-bar-test.js

```diff
@@ -105,110 +223,154 @@ describe('BucketBar', () => {
-    });
-
-    it('sets the tab text to the number of annotations', () => {
-      const tab = $('<div />');
-      const bucketBar = setup(tab);
-      bucketBar.buckets[0].push('Another Annotation?');
+    beforeEach(() => {
+      bucketBar = createBucketBar();
+      fakeAnchors = [
+        createAnchor(),
+        createAnchor(),
+        createAnchor(),
+        createAnchor(),
+        createAnchor(),
+        createAnchor(),
+      ];
+      // These two anchors are considered to be offscreen upwards
+      fakeAbove = [fakeAnchors[0], fakeAnchors[1]];
+      // These buckets are on-screen
+      fakeBuckets = [
+        { anchors: [fakeAnchors[2], fakeAnchors[3]], position: 350 },
+        { anchors: [], position: 450 }, // This is an empty bucket
+        { anchors: [fakeAnchors[4]], position: 550 },
+      ];
+      // This anchor is offscreen below
+      fakeBelow = [fakeAnchors[5]];
 
-      bucketBar._buildTabs();
-      assert.equal(tab.text(), '2');
+      fakeBucketUtil.constructPositionPoints.returns({
+        above: fakeAbove,
+        below: fakeBelow,
+        points: [],
+      });
+      fakeBucketUtil.buildBuckets.returns(fakeBuckets.slice());
     });
 
-    it('sets the tab text to the number of annotations', () => {
-      const tab = $('<div />');
-      const bucketBar = setup(tab);
-      bucketBar.buckets[0].push('Another Annotation?');
-
-      bucketBar._buildTabs();
-      assert.equal(tab.text(), '2');
-    });
+    describe('navigation bucket tabs', () => {
+      it('adds navigation tabs to scroll up and down to nearest anchors offscreen', () => {
+        bucketBar.update();
+        const validBuckets = nonEmptyBuckets(bucketBar);
+        assert.equal(
+          validBuckets[0].getAttribute('title'),
+          'Show 2 annotations'
+        );
+        assert.equal(
+          validBuckets[validBuckets.length - 1].getAttribute('title'),
+          'Show one annotation'
+        );
 
-    it('adds the class "upper" if the annotation is at the top', () => {
-      const tab = $('<div />');
-      const bucketBar = setup(tab);
-      sinon.stub(bucketBar, 'isUpper').returns(true);
+        assert.isTrue(validBuckets[0].classList.contains('upper'));
+        assert.isTrue(
+          validBuckets[validBuckets.length - 1].classList.contains('lower')
+        );
+      });
 
-      bucketBar._buildTabs();
-      assert.equal(tab.hasClass('upper'), true);
-    });
+      it('removes unneeded tab elements from the document', () => {
+        bucketBar.update();
+        const extraEl = document.createElement('div');
+        extraEl.className = 'extraTab';
+        bucketBar.element.append(extraEl);
+        bucketBar.tabs.push(extraEl);
+        assert.equal(bucketBar.tabs.length, bucketBar.buckets.length + 1);
+
+        // Resetting this return is necessary to return a fresh array reference
+        // on next update
+        fakeBucketUtil.buildBuckets.returns(fakeBuckets.slice());
```

Comment:
Oh, gawd, I lost like 90 minutes to this. Took me forever to figure out where the shared reference was that was causing the `BucketBar.buckets` array to grow and grow and grow across `update` calls. :facepalm:  And all to test a dumb implementation detail. But when I ran into troubles with this test, I feared there was a legit logic bug, so I'm glad I took the time to track it down.